resources/testharness.js

```diff
@@ -587,47 +587,55 @@ policies and contribution forms [3].
 
     function async_test(func, name, properties)
     {
-        if (tests.promise_setup_called) {
-            tests.status.status = tests.status.ERROR;
-            tests.status.message = '`async_test` invoked after `promise_setup`';
-            tests.complete();
-        }
         if (typeof func !== "function") {
             properties = name;
             name = func;
             func = null;
         }
         var test_name = get_test_name(func, name);
         var test_obj = new Test(test_name, properties);
-        if (func) {
-            var value = test_obj.step(func, test_obj, test_obj);
-
-            // Test authors sometimes return values to async_test, expecting us
-            // to handle the value somehow. Make doing so a harness error to be
-            // clear this is invalid, and point authors to promise_test if it
-            // may be appropriate.
-            //
-            // Note that we only perform this check on the initial function
-            // passed to async_test, not on any later steps - we haven't seen a
-            // consistent problem with those (and it's harder to check).
-            if (value !== undefined) {
-                var msg = 'Test named "' + test_name +
-                    '" passed a function to `async_test` that returned a value.';
-
-                try {
-                    if (value && typeof value.then === 'function') {
-                        msg += ' Consider using `promise_test` instead when ' +
-                            'using Promises or async/await.';
-                    }
-                } catch (err) {}
+        if (!func)
+            return test_obj;
+        
+        // If a promise_setup() defined earlier is still pending, defer the
+        // execution until it resolves.
+        var resume = async_test_execute.bind(null, func, test_obj, test_name);
+        if (tests.pending_async_tests !== null)
+            tests.pending_async_tests.push(resume);
+        else
+            resume();
 
-                tests.set_status(tests.status.ERROR, msg);
-                tests.complete();
-            }
-        }
         return test_obj;
     }
 
+    function async_test_execute(func, test_obj, test_name) {
```

Comment:
Couldn't this just be defined inside `async_test`?

Reply:
Done.